api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2776,27 +2998,333 @@ public void putObject(String bucketName, String objectName, InputStream stream,
+      KeyPair keypair) throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException,
+      IOException, InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+      InternalException, InvalidArgumentException, NoSuchPaddingException, IllegalBlockSizeException,
+      BadPaddingException, InvalidAlgorithmParameterException {
+
+    String keyEncryptionCipherMode = "RSA";
+    if (keypair == null) {
+      throw new InvalidArgumentException("empty key pair not allowed");
+    }
+    
+    putObject(bucketName, objectName, stream, size, contentType, keypair, keyEncryptionCipherMode);
+  }
+
+  /*
+   * Common method to encrypt data, encrypt keys and upload it to the server.
+   */
+  private void putObject(String bucketName, String objectName, InputStream stream, long size, String contentType,
+      Object key, String keyEncryptionCipherMode) throws NoSuchAlgorithmException, InvalidKeyException,
+      NoSuchPaddingException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException,
+      IOException, InvalidArgumentException, InvalidBucketNameException, InsufficientDataException, NoResponseException,
+      ErrorResponseException, InternalException, XmlPullParserException {
+
+    String dataEncryptionCipherMode = "AES/CBC/PKCS5Padding";
+    Key masterKey = null;
+
+    // Generate symmetric 256 bit AES key for data encryption
+    KeyGenerator symKeyGenerator = KeyGenerator.getInstance("AES");
+    symKeyGenerator.init(256);
+    SecretKey dataEncryptionKey = symKeyGenerator.generateKey();
+
+    // Generate an iv to be used for data encryption
+    byte[] iv = new byte[16];
+    SecureRandom ivSeed = new SecureRandom();
+    ivSeed.nextBytes(iv);
+
+    // Get CipherStream after encryption of input stream
+    CipherInputStream cipherInputStream = Crypto.encryptInputStream(stream, dataEncryptionKey, dataEncryptionCipherMode,
+        iv);
+
+    // get the master key type and typecast
+    if (key instanceof KeyPair) {
+      masterKey = ((KeyPair) key).getPublic();
+    } else if (key instanceof SecretKey) {
+      masterKey = (SecretKey) key;
+    } else {
+      throw new InvalidArgumentException("key should be an instance of KeyPair or SecretKey");
+    }
+
+    // encrypt the plain data key using master key
+    byte[] dataKey = dataEncryptionKey.getEncoded();
+    byte[] encryptedDataKey = Crypto.encrypt(dataKey, masterKey, keyEncryptionCipherMode);
+
+    // Prepare data to be put to the object header
+    String encDataKey = BaseEncoding.base64().encode(encryptedDataKey);
+    String ivString = BaseEncoding.base64().encode(iv);
+    
+    // Set the headermap with encryption related metadata
+    Map<String, String> headerMap = new HashMap<>();
+    headerMap.put("x-amz-meta-x-amz-key", encDataKey);
+    headerMap.put("x-amz-meta-x-amz-iv", ivString);
+    // matdesc is unused
+    headerMap.put("x-amz-meta-x-amz-matdesc", "{}");
```

Comment:
It's aws-sdk behavior kept as empty json.  @vadmeste can fill why is it purposefully needed.